ExtraEnemyCustomization/Customizations/EnemyAbilities/Abilities/CloakAbility.cs

```diff
@@ -0,0 +1,209 @@
+                    Agent.MovingCuller.Culler.AddRenderer(trsRenderer);
+                    _handlers.Add(new Handler()
+                    {
+                        hideShadowMode = Ability.HideShadow,
+                        minOpacity = Ability.CloakOpacity,
+                        originalRenderer = renderer,
+                        transitionRenderer = trsRenderer
+                    });
+                }
+            }
+        }
+
+        protected override void OnEnter()
+        {
+            _cloakState = State.Cloaking;
+            _startTime = Clock.Time;
+            _endTime = Clock.Time + Ability.CloakDuration;
+
+            if (Ability.DecloakAfterDelay >= 0.0f)
+            {
+                _decloakTimer = Clock.Time + Ability.DecloakAfterDelay;
+            }
+
+            if (Ability.RequireTagForDetectionWhileCloaking)
+            {
+                Agent.RequireTagForDetection = true;
+            }
+        }
+
+        protected override void OnExit()
+        {
+            _cloakState = State.Decloaking;
+            _startTime = Clock.Time;
+            _endTime = Clock.Time + Ability.DecloakDuration;
+            Agent.RequireTagForDetection = _reqTagOriginal;
+        }
+
+        protected override void OnUpdate()
+        {
+            float progress;
+            switch (_cloakState)
+            {
+                case State.Cloaking:
+                    progress = GetProgress();
+                    _handlers.ForEach(h => h.SetProgress(1.0f - progress));
+
+                    if (progress >= 1.0f)
+                    {
+                        _cloakState = State.None;
+                        Agent.MovingCuller.m_disableAnimatorCullingWhenRenderingShadow = true;
+                        Agent.SetAnimatorCullingEnabled(false);
+                    }
+                    break;
+
+                case State.Decloaking:
+                    progress = GetProgress();
+                    _handlers.ForEach(h => h.SetProgress(progress));
+
+                    if (progress >= 1.0f)
+                    {
+                        _cloakState = State.None;
+                        Agent.MovingCuller.m_disableAnimatorCullingWhenRenderingShadow = false;
+                        Agent.SetAnimatorCullingEnabled(true);
+                    }
+                    break;
+            }
+
+            if (_decloakTimer <= Clock.Time && Executing)
+            {
+                _decloakTimer = float.MaxValue;
+                DoExit();
+            }
+        }
+
+        protected override void OnAbilityLazyUpdate()
+        {
+            Agent.MovingCuller.Culler.NeedsShadowRefresh = true;
+        }
+
+        private float GetProgress()
+        {
+            return Mathf.InverseLerp(_startTime, _endTime, Clock.Time);
+        }
+
+        public enum State
+        {
+            None,
+            Cloaking,
+            Decloaking
+        }
+
+        public struct Handler
```

Comment:
You should probably check to make sure the renderers aren't null when you use them, as some random script could come in and do a little deleting

Reply:
I'll just lock the Handler type to private

Reply:
Just noticed what you meaning, I'll add null check and error log